edb/edgeql/compiler/func.py

```diff
@@ -179,7 +181,64 @@ def compile_operator(
 
         args.append((arg_type, arg_ir))
 
-    matched = polyres.find_callable(opers, args=args, kwargs={}, ctx=ctx)
+    matched = None
+    # Some 2-operand operators are special when their operands are
+    # arrays or tuples.
+    if len(args) == 2:
+        coll_opers = None
+        # If both of the args are arrays or tuples, potentially
+        # compile the operator for them differently than for other
+        # combinations.
+        if args[0][0].is_tuple() and args[1][0].is_tuple():
+            # Out of the candidate operators, find the ones that
+            # correspond to tuples.
+            coll_opers = [op for op in opers
+                          if all(param.get_type(schema).is_tuple() for param
+                                 in op.get_params(schema).objects(schema))]
+
+        elif args[0][0].is_array() and args[1][0].is_array():
+            # Out of the candidate operators, find the ones that
+            # correspond to arrays.
+            coll_opers = [op for op in opers
+                          if all(param.get_type(schema).is_array() for param
+                                 in op.get_params(schema).objects(schema))]
+
+        # Proceed only if we have a special case of collection operators.
+        if coll_opers is not None:
+            # Then check if they are recursive (i.e. validation must be
+            # done recursively for the subtypes). We rely on the fact that
+            # it is forbidden to define an operator that has both
+            # recursive and non-recursive versions.
+            if not coll_opers[0].get_recursive(schema):
+                # The operator is non-recursive, so regular processing
+                # is needed.
+                matched = polyres.find_callable(
+                    coll_opers, args=args, kwargs={}, ctx=ctx)
+
+            else:
+                # Ultimately the operator will be the same, regardless of the
+                # specific operand types, as long as it passed validation, so
+                # we just use the first operand type for the purpose of
+                # finding the callable.
+                matched = polyres.find_callable(
+                    coll_opers,
+                    args=[(args[0][0], args[0][1]), (args[0][0], args[1][1])],
+                    kwargs={}, ctx=ctx)
+
+                # Now that we have an operator, we need to validate that it
+                # can be applied to the tuple or array elements.
+                submatched = validate_recursive_operator(
+                    opers, args[0], args[1], ctx=ctx)
+
+                if len(submatched) != 1:
+                    # generate the error message corresponding to the number
+                    # of matches
+                    matched = submatched
```

Comment:
I think the comment doesn't match the code, no?

Reply:
I've reworded the comment to clarify what I mean.